snuba/query/matcher.py

```diff
@@ -0,0 +1,218 @@
+            strings={**self.strings, **values.strings},
+            ints={**self.ints, **values.ints},
+            floats={**self.floats, **values.floats},
+            dates={**self.dates, **values.dates},
+            datetimes={**self.datetimes, **values.datetimes},
+        )
+
+
+class Pattern(ABC, Generic[TNode]):
+    """
+    Tries to match a given node (like an AST Expression) with the rules provided
+    when instantiating this class. This is meant to work like a regular expression
+    but for AST expressions.
+
+    When a node is provided to this method, if this method return an instance of
+    MatchResult it means the node fully matches the pattern.
+    If the pattern defines named parameters, their value is returned as part of
+    the MatchResult object returned.
+    More on how to define parameters in the Param class.
+
+    This is not supposed to ever throw as long as the node is properly is a valid
+    data structure, even if it does not match the pattern.
+    """
+
+    @abstractmethod
+    def match(self, node: TNode) -> Optional[MatchResult]:
+        """
+        Returns a MatchResult if the node provided matches this pattern
+        otherwise it returns None.
+        """
+        raise NotImplementedError
+
+
+@dataclass(frozen=True)
+class Param(Pattern[TNode], Generic[TNode]):
+    """
+    Defines a named parameter in a Pattern. When matching the overall Pattern,
+    if the Pattern nested in this object matches, the corresponding input node
+    will be given a name and returned as part of the MatchResult return value.
+    This is meant to represent what a group is in a regex.
+
+    Params can be nested, and they can wrap any valid Pattern.
+    """
+
+    name: str
+    pattern: Pattern[TNode]
+
+    def match(self, node: TNode) -> Optional[MatchResult]:
+        result = self.pattern.match(node)
+        if not result:
+            return None
+
+        return result.merge(
+            MatchResult(
+                expressions={self.name: node} if isinstance(node, Expression) else {},
+                strings={self.name: node} if isinstance(node, str) else {},
+                ints={self.name: node} if isinstance(node, int) else {},
+                floats={self.name: node} if isinstance(node, float) else {},
+                dates={self.name: node} if isinstance(node, date) else {},
+                datetimes={self.name: node} if isinstance(node, datetime) else {},
+            )
+        )
+
+
+@dataclass(frozen=True)
+class AnyExpression(Pattern[Expression]):
+    """
+    Successfully matches any expression
+    """
+
+    def match(self, node: Expression) -> Optional[MatchResult]:
+        return MatchResult()
+
+
+@dataclass(frozen=True)
+class AnyString(Pattern[Optional[str]]):
+    """
+    Successfully matches any string
+    """
+
+    def match(self, node: Optional[str]) -> Optional[MatchResult]:
+        return MatchResult()
```

Comment:
Couldn't this just be `Any(Pattern[TNode])` since there are no runtime assertions occurring?

Reply:
It does not cover all cases. The value of a Literal for example is a Union[str, int, datetime, many others]. Applying Any() to the value of a literal would not yield the same result as AnyString(). 
I think they should both exist (the generic Any and AnyString if we want specifically a string, or others)

Reply:
I feel like I'm either reading the code or the comment incorrectly here. I don't see the difference between them other than the bound type variable, so why not leave it unbound so the implementation can be shared? `AnyString` isn't acting on `Literal` types anyway — just normal Python strings — so I don't see what that has to do with these.

Reply:
This is the example I am talking about:
```
exp = Literal(None, "bla")
exp2 = Literal(None, 123)

# Yes I know I did not create LiteralPattern yet, it would work like for Column
p = LiteralPattern(None, AnyString())
p2 = LiteralPattern(None, Any())

I would like p to match exp and not exp2
while p2 would match both.
```

Is this the way Any() would work or am I missing something ?

Reply:
Both `Any[str]` (as suggested) and `AnyString` (as implemented) would match both conditions at runtime since they don't use `isinstance`.

Reply:
I am dumb sorry about that, I wrote this some time ago and I did not remember what it was meant to do. Yes, in order to match literal values of specific types it would need to be different.

Reply:
This seems to do what we both want here:

```python

@dataclass(frozen=True)
class Any(Pattern[TNode]):
    """
    Successfully matches any expression
    """

    type: Type[TNode]

    def match(self, node: TNode) -> Optional[MatchResult]:
        return MatchResult() if isinstance(node, self.type) else None


m1: Any[None] = Any(type(None))
assert m1.match(None) is not None
assert m1.match("foo") is None  # also does not type check

m2: Any[str] = Any(str)
assert m2.match(None) is None
assert m2.match("foo") is not None  # also does not type check
```

Reply:
I suspect something similar would also work for things like `String` below (preventing a lot of boilerplate matchers for simple types) as long as the type bound includes `__eq__`, kind of similar to how type bounding `Interval` works here: https://github.com/getsentry/snuba/blob/e4a21104908c736f8dcbd24a682bfdb62f4187bd/snuba/utils/types.py#L7-L67

Reply:
Unfortunately I think `Any(Pattern[TNode])` still does not cover all cases:
1) we cannot use to match any expression `Any(Expression)`because Expression is an abstract class. `Only concrete class can be given where "Type[Expression]" is expectedmypy(error)`

2) It cannot match something like an Optional[str] which is going to be common to extract for example the table name or the alias from a column, there you want to match a nullable string. Which technically you can achieve with `Or([Any(type(None)), Any(str)])` though at that point I would argue `AnyString()` is a cleaner and easier option.

Reply:
> 1. we cannot use it to match "any" expression `Any(Expression)` because Expression is an abstract class. `Only concrete class can be given where "Type[Expression]" is expectedmypy(error)`

Hm, weird. Okay.

> 2. It cannot match something like an Optional[str] which is going to be common to extract for example the table name or the alias from a column, there you want to match a nullable string. Which technically you can achieve with `Or([Any(type(None)), Any(str)])` though at that point I would argue `AnyString()` is a cleaner and easier option.

I actually think that `AnyString` matching `None` is a little misleading personally given that `None` is not a string.

Reply:
Fair, so what about we require `Or([Any(type(None)), Any(str)]` to extract the alias of an arbitrary expression and revisit if/when it becomes a pain, adding a new Pattern type will be a trivial task.

Reply:
Or `AnyOptionalString` ?

Reply:
I think either of those approaches make sense but the wording "arbitrary expression" is confusing me a bit here.

It's super frustrating that `Union` (and subsequently `Optional`) can't be used with `isinstance` because that would simplify things a lot here. I'm also assuming that runtime type checking actually is a necessity back based on the prior conversation.

Thinking about that further, all of these patterns that require type sensitivity/awareness have no guarantee that the type that is being provided to them matches the parameter type in the static definition ([as in the example above](https://github.com/getsentry/snuba/pull/942#discussion_r427609554)), which itself is probably an issue.

For general purpose uses, shouldn't the `node` parameter type here something looser — possibly even `Any`, but at least something that represents the set of all types that may be present in a query, roughly what is currently in `MatchResult` — since there are no guarantees what type is going to be passed into the pattern?

Reply:
> Thinking about that further, all of these patterns that require type sensitivity/awareness have no guarantee that the type that is being provided to them matches the parameter type in the static definition (as in the example above), which itself is probably an issue.

Are you talking specifically about the Any Pattern or in general for Patterns (like a FunctionCall pattern guarantees that it receives at least an Expression bjt I don't believe this is specifically an issue also because I doubt we could achieve anything better). I think the value of those static types is in composing the Pattern expression. Like you cannot try to use an Any(str) or a String("asd") to match a field that, in the AST would be an Expression and this is statically guaranteed.

> For general purpose uses, shouldn't the node parameter type here something looser — possibly even Any, but at least something that represents the set of all types that may be present in a query, roughly what is currently in MatchResult — since there are no guarantees what type is going to be passed into the pattern?

In order to do that we would have to drop the static checking we do when composing the Pattern. If Any was matching anything, it could not be used to match the name of a function (for example) because that we know, in the AST, is a string, thus we enforce that if you try to match a FunctionCall, the Pattern you pass to match the function Name must be a Pattern[String]. 
I believe there is a lot of value in preserving strict type checking when building a Pattern because it makes it a lot harder to write impossible Patterns, so I would rather not compromise on that side.

Reply:
I feel like there are two slightly different goals here, and I was only addressing one of them. Here is my attempt to make those goals explicit:

1. It is helpful to have a pattern matcher that matches values of a known type that is statically defined, and the value is validated/enforced at runtime (in some other part of the program) to be of the correct type if it is user provided data. To use your example, the function name in a function call can only be a string, and we can be confident that it actually _is_ a string at runtime, since a.) query parser enforces it as such for user queries and, b.) we have static type definitions for things like processors that may add their own functions. Runtime checks are unnecessary here as long as the parser is holding up it's end of the bargain and we're diligent with static typing.
2. It is helpful to have a pattern matcher that matches `Literal.value` the value is of specific type. Using the [example above again](https://github.com/getsentry/snuba/pull/942#discussion_r427609554), the type of the `node` parameter could be any of [these possible value types](https://github.com/getsentry/snuba/blob/0959cbd98ce91f7b5b19a285208efb60ca9ff38d/snuba/query/expressions.py#L115), which _require_ the value be checked at runtime to see if the pattern matches. Any pattern matcher that operates on a specific subtype of `Expression` has the same constraint (and `Column`, `FunctionCall`, etc.) are already doing this for that reason.)

Is this a fair summarization?

Reply:
I think it misses one goal (which actually was to me the biggest driver to have the Pattern class to be gneric in the first place). 
It is helpful that we can make it hard statically to write a nonsensical Pattern that could never match anything because it does not comply with the AST structure itself. 
Example: statically prevent someone to try to match a function parameter with a datetime. A function parameter can only be an expression, so it should not be possible to write a pattern like:
```
p = FunctionCall(
      Param("alias", String("asdasd")),     # OK
      String("f_name"), # OK
      (
                Param("p_1", Datetime(some valid date)), # -> NOT OK a parameter cannot be a datetime object
            ),
        ),
```
In the AST a parameter of a FunctionCall is an Expression, it would be great if it we could statically guarantee that you cannot write such a Pattern.